feat(core): make signals interoperable

divdavem opened the pull request:
Hello,

I am opening this PR to suggest making Angular signals interoperable with other signal libraries by implementing what I suggested [in this ticket of the TC39 signals proposal](https://github.com/tc39/proposal-signals/issues/259).

This PR contains an implementation of my suggestion. This is a work in progress. As mentioned in the above ticket, it would be good to receive feedback from different signal libraries authors, so this PR may evolve as feedback is received.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

The following tests (and many similar tests) currently fail:

```ts
import { computed as tansuComputed, writable as tansuSignal } from '@amadeus-it-group/tansu';
import { computed as angularComputed, signal as angularSignal } from '@angular/core';
import { expect, it } from 'vitest';

it('should be interoperable from tansu to angular', () => {
  const a = tansuSignal(0);
  const c = angularComputed(() => a() * 2);
  expect(c()).toEqual(0);
  a.set(1);
  expect(c()).toEqual(2);
});

it('should be interoperable from angular to tansu', () => {
  const a = angularSignal(0);
  const c = tansuComputed(() => a() * 2);
  expect(c()).toEqual(0);
  a.set(1);
  expect(c()).toEqual(2);
});

```

It is also related to #50189

## What is the new behavior?

The above tests succeed.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

This PR can break code that directly depends on `@angular/core/primitives/signals` because of changes in the data types.

## Other information

Note that the content of the `interop_lib.ts` file in this PR is directly copied from [this ticket](https://github.com/tc39/proposal-signals/issues/259) as a temporary solution. It should ideally come from an independent npm package that provides the interoperability layer.

divdavem committed:
feat(core): make signals interoperable

divdavem committed:
chore: "public-api:update" and "symbol-extractor:update"

thePunderWoman commented:
@divdavem We appreciate the effort here, but at the moment things are moving too quickly with signals to worry about interop. Once we see signals move further along and as our internal signals development work slows, we can worry about interop at that time. Thank you for your contribution though! We appreciate you!

divdavem commented:
@thePunderWoman Thank you for your answer
I am happy to finally see someone commenting on my PR, even though I am disappointed that it is only to (kindly) close it.
I was not necessarily expecting to have my PR merged soon, but I mostly opened this PR to have some feedback from the Angular team regarding my suggested interop specification. I understand that the Angular team is busy.

Note that the code I implemented in this PR no longer reflects my latest [interop specification](https://github.com/divdavem/tansu/blob/interopSignal/src/interop.ts). (I updated it following some feedback given by another signal library implementation author during a TC 39 Signal Community meeting).
I am happy to open another PR on Angular to implement my latest interop specification when when I have time (and if there is some interest from the Angular team/community).
